File: lib/KalmanFilter.cpp
#include "KalmanFilter.h"
#include <optional>

namespace ctrl
{

    KalmanFilter::KalmanFilter(const StateSpace &plant,
                               const Eigen::MatrixXd &Q_noise,
                               const Eigen::MatrixXd &R_noise,
                               const Eigen::MatrixXd &P0)
        : plant_(plant), Q_(Q_noise), R_(R_noise), Ts_(plant.Ts)
    {
        const int n = plant.stateSize();
        const int p = plant.outputSize();
        x_hat_ = Eigen::VectorXd::Zero(n);
        P_ = P0.rows() == n ? P0 : Eigen::MatrixXd::Identity(n, n);
        R_safe_.resize(p, p);
        S_    .resize(p, p);
        Kf_   .resize(n, p);
        IKC_  .resize(n, n);
        P_new_.resize(n, n);
    }

    // Predict: advance state estimate and inflate covariance with process noise.
    void KalmanFilter::predict(const Eigen::VectorXd &u)
    {
        x_hat_ = plant_.A * x_hat_ + plant_.B * u;
        P_ = plant_.A * P_ * plant_.A.transpose() + Q_;
    }

    // Update: incorporate measurement y[k], correct estimate, deflate covariance.
    // Joseph form for P update maintains positive semi-definiteness numerically.
    void KalmanFilter::update(const Eigen::VectorXd &y,
                              const Eigen::VectorXd &u_current)
    {
        const Eigen::MatrixXd &C = plant_.C;
        const Eigen::MatrixXd &D = plant_.D;
        const int p = plant_.outputSize();
        const int n = plant_.stateSize();

        // Enforce minimum noise floor on R_ into pre-allocated R_safe_
        R_safe_ = R_;
        R_safe_.diagonal() = R_safe_.diagonal().cwiseMax(1e-12);

        // Innovation covariance (pre-allocated S_)
        S_.noalias() = C * P_ * C.transpose();
        S_ += R_safe_;

        // Kalman gain (pre-allocated Kf_) - skip update entirely if S is numerically singular
        const auto ldlt = S_.ldlt();
        if (ldlt.info() != Eigen::Success)
            return;
        Kf_.noalias() = P_ * C.transpose() * ldlt.solve(Eigen::MatrixXd::Identity(p, p));

        // State update
        const Eigen::VectorXd innov = y - C * x_hat_ - D * u_current;
        if (mismatch_det_) mismatch_det_->update(innov);
        x_hat_.noalias() += Kf_ * innov;

        // Covariance update - Joseph form: P = (I-KC).P.(I-KC)' + K.R.K'
        // Accumulate into P_new_ to avoid aliasing (P_ appears on both sides)
        IKC_.noalias() = Eigen::MatrixXd::Identity(n, n) - Kf_ * C;
        P_new_.noalias() = IKC_ * P_ * IKC_.transpose();
        P_new_.noalias() += Kf_ * R_safe_ * Kf_.transpose();
        P_ = P_new_;
    }

    void KalmanFilter::step(const Eigen::VectorXd &y,
                            const Eigen::VectorXd &u_prev,
                            std::optional<std::reference_wrapper<const Eigen::VectorXd>> u_current)
    {
        predict(u_prev);
        // Use u_current for the D.u innovation term when provided; fall back to u_prev for D=0 plants.
        update(y, u_current.has_value() ? u_current->get() : u_prev);
    }

    // Plain-reference overload - avoids std::optional<std::reference_wrapper<...>> for pybind11.
    void KalmanFilter::step(const Eigen::VectorXd &y,
                            const Eigen::VectorXd &u_prev,
                            const Eigen::VectorXd &u_current)
    {
        predict(u_prev);
        update(y, u_current);
    }

    void KalmanFilter::reset()
    {
        x_hat_.setZero();
        P_ = Eigen::MatrixXd::Identity(plant_.stateSize(), plant_.stateSize());
    }

} // namespace ctrl

File: MATLAB/+ctrl/KalmanFilter.m
classdef KalmanFilter < handle
%KALMANFILTER  Discrete-time linear Kalman filter.  Port of ctrl::KalmanFilter.
%
%       x[k+1] = A x[k] + B u[k] + w,   w ~ N(0, Q)
%       y[k]   = C x[k] + D u[k] + v,   v ~ N(0, R)
%
%   Construction:
%       kf = ctrl.KalmanFilter(plant, Q_noise, R_noise)
%       kf = ctrl.KalmanFilter(plant, Q_noise, R_noise, P0)   % P0 defaults to I
%   where plant is a ctrl.StateSpace.
%
%   Predict/update use the Joseph form for the covariance (numerically PSD-preserving),
%   and skip the update if the innovation covariance is not positive definite.
%   Equivalent MATLAB (Control System Toolbox): kalman()/kalmd().

  properties
    A; B; C; D
    Q; R
    xhat   % state estimate x^[k|k]
    P      % error covariance P[k|k]
    Ts
    n; p
  end

  methods
    function obj = KalmanFilter(plant, Q_noise, R_noise, P0)
      if nargin < 4, P0 = []; end
      obj.A = plant.A; obj.B = plant.B; obj.C = plant.C; obj.D = plant.D;
      obj.Q = Q_noise; obj.R = R_noise; obj.Ts = plant.Ts;
      obj.n = size(plant.A, 1);
      obj.p = size(plant.C, 1);
      obj.xhat = zeros(obj.n, 1);
      if ~isempty(P0) && size(P0, 1) == obj.n
        obj.P = P0;
      else
        obj.P = eye(obj.n);
      end
    end

    function predict(obj, u)
      %PREDICT  Advance the estimate with the previous control input u[k-1].
      obj.xhat = obj.A * obj.xhat + obj.B * u(:);
      obj.P    = obj.A * obj.P * obj.A.' + obj.Q;
    end

    function update(obj, y, u_current)
      %UPDATE  Incorporate measurement y[k] (u_current used for the D u feedthrough).
      Rsafe = obj.R;
      for i = 1:size(Rsafe, 1)           % minimum noise floor on the diagonal
        Rsafe(i, i) = max(Rsafe(i, i), 1e-12);
      end

      S = obj.C * obj.P * obj.C.' + Rsafe;

      [~, pflag] = chol(S);              % skip update if S is not positive definite
      if pflag ~= 0
        return;
      end

      Kf    = (obj.P * obj.C.') / S;
      innov = y(:) - obj.C * obj.xhat - obj.D * u_current(:);
      obj.xhat = obj.xhat + Kf * innov;

      % Joseph form: P = (I-KC) P (I-KC)' + K R K'
      IKC   = eye(obj.n) - Kf * obj.C;
      obj.P = IKC * obj.P * IKC.' + Kf * Rsafe * Kf.';
    end

    function step(obj, y, u_prev, u_current)
      %STEP  predict(u_prev) then update(y, u_current).  u_current defaults to u_prev.
      if nargin < 4, u_current = u_prev; end
      obj.predict(u_prev);
      obj.update(y, u_current);
    end

    function reset(obj)
      obj.xhat = zeros(obj.n, 1);
      obj.P    = eye(obj.n);
    end

    function x = state(obj),      x = obj.xhat; end
    function P = covariance(obj), P = obj.P;    end
    function t = sampleTime(obj), t = obj.Ts;   end
  end
end
